src/Filecoin/Address.cpp

```diff
@@ -16,165 +15,283 @@ namespace TW::Filecoin {
-        }
-        if (data.size() == 11 && data[10] > 0x01) {
-            return false;
-        }
-        std::size_t i;
-        for (i = 1; i < data.size(); i++) {
-            if ((data[i] & 0x80) == 0) {
-                break;
-            }
-        }
-        return i == data.size() - 1;
-    } else {
-        return data.size() == (1 + Address::payloadSize(type));
     }
-}
 
-static bool isValidID(const std::string& string) {
-    if (string.length() > 22)
-        return false;
-    for (auto i = 2ul; i < string.length(); i++) {
-        if (string[i] < '0' || string[i] > '9') {
-            return false;
-        }
-    }
     try {
         size_t chars;
-        [[maybe_unused]] uint64_t id = std::stoull(string.substr(2), &chars);
+        actorID = std::stoull(string, &chars);
         return chars > 0;
     } catch (...) {
         return false;
     }
 }
 
-static bool isValidBase32(const std::string& string, Address::Type type) {
-    // Check if valid Base32.
-    uint8_t size = Address::payloadSize(type);
-    Data decoded;
-    if (!Base32::decode(string.substr(2), decoded, BASE32_ALPHABET_FILECOIN)) {
-        return false;
+/// Decodes the given `string` as an ActorID.
+/// Please note `bytes` must not contain any prefixes.
+/// https://github.com/paritytech/unsigned-varint/blob/a3a5b8f2bee1f44270629e96541adf805a53d32c/src/decode.rs#L66-L86
+static bool decodeActorID(const Data& bytes, uint64_t& actorID, std::size_t& remainingPos) {
+    const std::size_t maxBytes = 9;
+
+    actorID = 0;
+    remainingPos = 0;
+    for (remainingPos = 0; remainingPos < bytes.size() && remainingPos <= maxBytes; ++remainingPos) {
+        auto byte = bytes[remainingPos];
+        auto k = byte & SCHAR_MAX;
+        actorID |= k << (remainingPos * 7);
+
+        // Check if last.
+        if ((byte & 0x80) == 0) {
+            if (byte == 0 && remainingPos > 0) {
+                // If last byte is zero, it could have been "more minimally" encoded by dropping that trailing zero.
+                return false;
+            }
+            ++remainingPos;
+            return true;
+        }
     }
 
-    // Check size
-    if (decoded.size() != size + checksumSize) {
-        return false;
+    // Couldn't find the last byte so `(byte & 0x80) == 0`.
+    return false;
+}
+
+static void writeActorID(uint64_t actorID, Data& dest) {
+    while (actorID >= 0x80) {
+        dest.push_back(((uint8_t)actorID) | 0x80);
+        actorID >>= 7;
     }
+    dest.push_back((uint8_t)actorID);
+}
 
-    // Extract raw address.
-    Data address;
-    address.push_back(static_cast<uint8_t>(type));
-    address.insert(address.end(), decoded.data(), decoded.data() + size);
+/// Returns encoded bytes of Address including the protocol byte and actorID (if required)
+/// without the checksum.
+static Data addressToBytes(Address::Type type, uint64_t actorID, const Data& payload) {
+    Data encoded;
+    encoded.push_back(static_cast<uint8_t>(type));
```

Comment:
same remark, if u use emplace_back, static_cast should not be required.

Reply:
In this case it's required, because the compiler doesn't know how to emplace `Address::Type` (even though it's `: uint8_t`) as `uint8_t`.